src/sentry/db/models/base.py

```diff
@@ -155,6 +155,74 @@ class Meta:
 
     __repr__ = sane_repr("id")
 
+    def save(self, *args: Any, **kwds: Any) -> None:
```

Comment:
So here's the trick.  Originally, I tried to implement model level outbox logic via an abstract mixin that override save.  The problem is that multiple inheritance makes overriding methods (esp save) very tricky since the ordering of the mixin changes the order of the super, which alters a lot of behavior.

To have automatic saving of outboxes, I decided that putting the logic in the one common super class was better than hoping that a mixin is ordered correctly in the MRO (remember, I can't *enforce* the MRO because `BaseModel` uses a metaclass, and I can't check the MRO at runtime unless I can be the metaclass, but python doesn't let me have multiple metaclass soo....).  Basically, as weird as it looks, I decided that the simplest implementation that was still really clear and easy to understand is to put the outbox behavior into the base model, and use super classes to *signal* that behavior, rather than carry it.

I'm open to counter opinions.

Reply:
Seems reasonable. Maybe put a note in the `Produces*OutboxOnUpdate` docstrings indicating that `Model` looks for them as a signal.